{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "65c215b5-4e19-410f-ba34-24b9d96291ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0bbf8c45-e50d-43bd-b58d-3497e70c1d45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read scRNA-seq data\n",
    "adata = sc.read_h5ad(\"SEAAD_MTG_RNAseq_final-nuclei.2024-02-13.h5ad\", backed='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dce05d20-419d-471e-af0d-a0b8ebdecff3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['sample_id', 'Neurotypical reference', 'Donor ID', 'Organism',\n",
      "       'Brain Region', 'Sex', 'Gender', 'Age at Death', 'Race (choice=White)',\n",
      "       'Race (choice=Black/ African American)',\n",
      "       ...\n",
      "       'Used in analysis', 'Class confidence', 'Class', 'Subclass confidence',\n",
      "       'Subclass', 'Supertype confidence', 'Supertype (non-expanded)',\n",
      "       'Supertype', 'Continuous Pseudo-progression Score',\n",
      "       'Severely Affected Donor'],\n",
      "      dtype='object', length=135)\n"
     ]
    }
   ],
   "source": [
    "print(adata.obs.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e7e8c02-05c3-469e-bfab-6a9bc15e7853",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Human MTG', 'Human MTG_L5', 'Human MTG All Layers']\n",
      "Categories (3, object): ['Human MTG', 'Human MTG All Layers', 'Human MTG_L5']\n"
     ]
    }
   ],
   "source": [
    "print(adata.obs[\"Brain Region\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae6aadd4-60f9-4cd8-a617-b0e58bc85e2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total cells: 1378211\n",
      "Total genes: 36601\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total cells: {adata.n_obs}\")\n",
    "print(f\"Total genes: {adata.n_vars}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c5520897-22e8-4bb0-91b1-62412bc5a99c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cells: 34922\n"
     ]
    }
   ],
   "source": [
    "mask = adata.obs[\"Brain Region\"] == \"Human MTG_L5\"\n",
    "selected_cells = adata.obs_names[mask]\n",
    "\n",
    "print(\"Number of cells:\", len(selected_cells))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "693df787-e8fb-487b-80be-d5623fb84422",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_MTG_L5 = sc.read_h5ad(\"SEAAD_MTG_RNAseq_final-nuclei.2024-02-13.h5ad\", backed='r')[selected_cells, :].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba1d16cd-d620-430c-81b1-fe85911805be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_id', 'Neurotypical reference', 'Donor ID', 'Organism',\n",
       "       'Brain Region', 'Sex', 'Gender', 'Age at Death', 'Race (choice=White)',\n",
       "       'Race (choice=Black/ African American)',\n",
       "       ...\n",
       "       'Used in analysis', 'Class confidence', 'Class', 'Subclass confidence',\n",
       "       'Subclass', 'Supertype confidence', 'Supertype (non-expanded)',\n",
       "       'Supertype', 'Continuous Pseudo-progression Score',\n",
       "       'Severely Affected Donor'],\n",
       "      dtype='object', length=135)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_MTG_L5.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0eb027e5-fc7e-4727-acc7-a910bbd601ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Subclass\n",
       "L5 IT              8807\n",
       "L4 IT              4613\n",
       "Sst                4395\n",
       "Pvalb              4336\n",
       "L6 IT Car3         3068\n",
       "L6 IT              1787\n",
       "L5/6 NP            1620\n",
       "Vip                1403\n",
       "L6b                1120\n",
       "L2/3 IT            1049\n",
       "L6 CT               840\n",
       "Lamp5 Lhx6          830\n",
       "Lamp5               516\n",
       "L5 ET               216\n",
       "Chandelier          164\n",
       "Sst Chodl            41\n",
       "Oligodendrocyte      36\n",
       "Pax6                 30\n",
       "Endothelial          16\n",
       "Astrocyte            14\n",
       "Sncg                  7\n",
       "Microglia-PVM         7\n",
       "OPC                   6\n",
       "VLMC                  1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_MTG_L5.obs[\"Subclass\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0deda364-abc1-42f6-9122-7f14291488b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Class\n",
       "Neuronal: Glutamatergic        23120\n",
       "Neuronal: GABAergic            11722\n",
       "Non-neuronal and Non-neural       80\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_MTG_L5.obs[\"Class\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e790d899-95ef-4b0b-bec6-833255bede4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize and log-transform\n",
    "sc.pp.normalize_total(adata_MTG_L5, target_sum=1e4)\n",
    "sc.pp.log1p(adata_MTG_L5)\n",
    "\n",
    "# Select top 2000 highly variable genes (optional but recommended)\n",
    "sc.pp.highly_variable_genes(adata_MTG_L5, n_top_genes=2000, subset=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "87700838-a784-44db-8a5b-7e7906b98109",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(\n",
    "    adata_MTG_L5,\n",
    "    groupby=\"Subclass\",\n",
    "    groups=[\"L5 IT\"],     # cell type of interest\n",
    "    reference=\"rest\",     # compare vs all other subclasses\n",
    "    method=\"wilcoxon\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "838523ab-2ac5-4e17-ba94-6a219d403457",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IL1RAPL2</td>\n",
       "      <td>108.958481</td>\n",
       "      <td>2.702137</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASC15</td>\n",
       "      <td>108.690010</td>\n",
       "      <td>2.023123</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CPNE4</td>\n",
       "      <td>105.508072</td>\n",
       "      <td>1.810703</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LRRK1</td>\n",
       "      <td>100.498947</td>\n",
       "      <td>2.297281</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IPCEF1</td>\n",
       "      <td>100.253082</td>\n",
       "      <td>2.030481</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LDLRAD3</td>\n",
       "      <td>99.464058</td>\n",
       "      <td>2.827661</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HS3ST2</td>\n",
       "      <td>99.157211</td>\n",
       "      <td>2.084499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TLL1</td>\n",
       "      <td>97.960358</td>\n",
       "      <td>2.742297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RXFP1</td>\n",
       "      <td>96.707504</td>\n",
       "      <td>1.977844</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RORB</td>\n",
       "      <td>95.499580</td>\n",
       "      <td>2.749091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SORCS1</td>\n",
       "      <td>94.679375</td>\n",
       "      <td>1.959579</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FSTL4</td>\n",
       "      <td>93.315239</td>\n",
       "      <td>1.669531</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ANO3</td>\n",
       "      <td>91.500984</td>\n",
       "      <td>2.043158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EPHA3</td>\n",
       "      <td>90.988922</td>\n",
       "      <td>2.023094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CLSTN2</td>\n",
       "      <td>90.320770</td>\n",
       "      <td>1.465642</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AC109466.1</td>\n",
       "      <td>90.307289</td>\n",
       "      <td>2.435190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AC060765.2</td>\n",
       "      <td>90.287819</td>\n",
       "      <td>2.812883</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CADPS2</td>\n",
       "      <td>88.809784</td>\n",
       "      <td>1.724164</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AC008415.1</td>\n",
       "      <td>86.834152</td>\n",
       "      <td>2.598839</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>LY86-AS1</td>\n",
       "      <td>86.033028</td>\n",
       "      <td>2.042343</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         names      scores  logfoldchanges  pvals  pvals_adj\n",
       "0     IL1RAPL2  108.958481        2.702137    0.0        0.0\n",
       "1       CASC15  108.690010        2.023123    0.0        0.0\n",
       "2        CPNE4  105.508072        1.810703    0.0        0.0\n",
       "3        LRRK1  100.498947        2.297281    0.0        0.0\n",
       "4       IPCEF1  100.253082        2.030481    0.0        0.0\n",
       "5      LDLRAD3   99.464058        2.827661    0.0        0.0\n",
       "6       HS3ST2   99.157211        2.084499    0.0        0.0\n",
       "7         TLL1   97.960358        2.742297    0.0        0.0\n",
       "8        RXFP1   96.707504        1.977844    0.0        0.0\n",
       "9         RORB   95.499580        2.749091    0.0        0.0\n",
       "10      SORCS1   94.679375        1.959579    0.0        0.0\n",
       "11       FSTL4   93.315239        1.669531    0.0        0.0\n",
       "12        ANO3   91.500984        2.043158    0.0        0.0\n",
       "13       EPHA3   90.988922        2.023094    0.0        0.0\n",
       "14      CLSTN2   90.320770        1.465642    0.0        0.0\n",
       "15  AC109466.1   90.307289        2.435190    0.0        0.0\n",
       "16  AC060765.2   90.287819        2.812883    0.0        0.0\n",
       "17      CADPS2   88.809784        1.724164    0.0        0.0\n",
       "18  AC008415.1   86.834152        2.598839    0.0        0.0\n",
       "19    LY86-AS1   86.033028        2.042343    0.0        0.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# identify genes that are differentially expressed in L5 IT vs other cell types\n",
    "deg = sc.get.rank_genes_groups_df(adata_MTG_L5, group=\"L5 IT\")\n",
    "deg.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5d1ff108-a99f-4a01-8000-b5a480d55443",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Open in backed mode to avoid loading everything, read scATAC-seq data\n",
    "adata = sc.read_h5ad(\"SEAAD_MTG_ATACseq_final-nuclei.2024-12-06.h5ad\", backed=\"r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2c2ec7e4-28b1-4ae2-b924-96133e661cf3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['sample_id', 'Neurotypical reference', 'Donor ID', 'Organism',\n",
      "       'Brain Region', 'Sex', 'Gender', 'Age at Death', 'Race (choice=White)',\n",
      "       'Race (choice=Black/ African American)',\n",
      "       ...\n",
      "       'Class', 'Subclass confidence', 'Subclass', 'Supertype confidence',\n",
      "       'Supertype (non-expanded)', 'Supertype', 'RNA Quality Control Score',\n",
      "       'Quality Control Clusters', 'Continuous Pseudo-progression Score',\n",
      "       'Severely Affected Donor'],\n",
      "      dtype='object', length=129)\n"
     ]
    }
   ],
   "source": [
    "print(adata.obs.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c2e03fa1-112d-4204-ab4f-66467db09d36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total cells: 516389\n",
      "Total peaks: 218882\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total cells: {adata.n_obs}\")\n",
    "print(f\"Total peaks: {adata.n_vars}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5c00702f-4041-453e-8321-736a65fa8dc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Subclass\n",
       "L2/3 IT            143333\n",
       "L4 IT               65159\n",
       "L5 IT               48096\n",
       "Vip                 36557\n",
       "Pvalb               36282\n",
       "Oligodendrocyte     35233\n",
       "Sst                 26008\n",
       "Astrocyte           19636\n",
       "L6 IT               18617\n",
       "Lamp5               16411\n",
       "L6 IT Car3          10021\n",
       "Microglia-PVM        9663\n",
       "OPC                  8563\n",
       "Sncg                 8413\n",
       "Lamp5 Lhx6           7660\n",
       "L5/6 NP              7187\n",
       "L6 CT                5797\n",
       "L6b                  5001\n",
       "Pax6                 2884\n",
       "Chandelier           2827\n",
       "VLMC                 1261\n",
       "L5 ET                 993\n",
       "Sst Chodl             488\n",
       "Endothelial           299\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"Subclass\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "23df5e5e-da92-4a85-a0b2-da8672bd2e4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import snapatac2 as snap\n",
    "# run pip install snapatac2 if not install yet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a6fc9685-4e66-47bb-a68b-57eeb18eb430",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [chr4:164130572-164131513, chr4:16412872-16415576, chr4:164116366-164116739, chr4:164083047-164083256, chr4:164056041-164057013]\n"
     ]
    }
   ],
   "source": [
    "print(adata.var.head()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e9c0fd4f-3a1d-4c13-bf90-94248707b7ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c7757ef6-6cf4-4ef5-86a3-9ecdb5e9e239",
   "metadata": {},
   "outputs": [],
   "source": [
    "target = \"Astrocyte\"\n",
    "group_col = \"Subclass\"\n",
    "\n",
    "mask_target = (adata.obs[group_col] == target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1841285c-0bee-41b2-b0a3-c98a3f430254",
   "metadata": {},
   "outputs": [],
   "source": [
    "# building a background set of random cells from all other cell types except the target cell type\n",
    "barcodes = np.array(adata.obs_names)\n",
    "background = []\n",
    "\n",
    "for ct in adata.obs[group_col].unique():\n",
    "    if ct == target:\n",
    "        continue\n",
    "    mask = (adata.obs[group_col] == ct)\n",
    "    n = mask.sum()\n",
    "    if n > 0:\n",
    "        k = min(50, n)  # up to 50 per cell type\n",
    "        background.append(np.random.choice(barcodes[mask], size=k, replace=False))\n",
    "\n",
    "background = np.concatenate(background)\n",
    "mask_background = np.isin(adata.obs_names, background)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d80ec1ec-4298-49ac-b8f0-e500a394d3b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-20 23:09:17 - INFO - Input contains 218882 features, now perform filtering with 'min_log_fc = 0.25' and 'min_pct = 0.05' ...\n",
      "2025-11-20 23:09:52 - INFO - Testing 30584 features ...\n",
      " 40%|████      | 12263/30584 [12:59<18:44, 16.29it/s]"
     ]
    }
   ],
   "source": [
    "# identify open chromatin regions that are selectively enriched in astrocyte, this will takes a while\n",
    "da = snap.tl.diff_test(\n",
    "    adata,\n",
    "    cell_group1 = mask_target,\n",
    "    cell_group2 = mask_background,\n",
    "    direction = \"positive\"     # enriched in cell_group1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4081e620-8d3c-4a60-a822-ae42ac6bef28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to get results returned by SnapATAC2\n",
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c8d8e804-9c9b-42b4-b4ec-b60255d59e25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape: (5, 4)\n",
      "┌──────────────────────────┬───────────────────┬─────────┬──────────────────┐\n",
      "│ feature name             ┆ log2(fold_change) ┆ p-value ┆ adjusted p-value │\n",
      "│ ---                      ┆ ---               ┆ ---     ┆ ---              │\n",
      "│ str                      ┆ f32               ┆ f64     ┆ f64              │\n",
      "╞══════════════════════════╪═══════════════════╪═════════╪══════════════════╡\n",
      "│ chr4:1789181-1815273     ┆ 2.38265           ┆ 0.0     ┆ 0.0              │\n",
      "│ chr4:1775512-1781064     ┆ 3.34286           ┆ 0.0     ┆ 0.0              │\n",
      "│ chr4:2785634-2802095     ┆ 2.295867          ┆ 0.0     ┆ 0.0              │\n",
      "│ chr5:38465985-38470039   ┆ 3.489634          ┆ 0.0     ┆ 0.0              │\n",
      "│ chr5:139805058-139809732 ┆ 3.697418          ┆ 0.0     ┆ 0.0              │\n",
      "└──────────────────────────┴───────────────────┴─────────┴──────────────────┘\n"
     ]
    }
   ],
   "source": [
    "# keep only statistically significant peaks\n",
    "da_sig = da.filter(pl.col(\"adjusted p-value\") < 0.01)\n",
    "\n",
    "print(da_sig.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6713578-c3cd-4c1b-b23a-48c29c96d00d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
